### DISCUSSION TITLE
I can render my Quarto project in R and see the interactive plot_ly graph, but the graph does not appear when I publish to Quarto Pubs.

### DISCUSSION THREAD

**Discussion by curtisld** [quarto-dev/quarto-cli#11175] [Q&A] [answered]
### Description

I've been able to render a Quarto file in RStudio with an interactive plot_ly graph (Run Document), but the html file produced has a blank space where the graph should be.  Perhaps not surprisingly, when I publish to Quarto Pubs there is still no embedded graph. 

I was able to publish to shinyapps.io and the interactive graph was included (see: https://curtisld.shinyapps.io/nla_explore/). 

The project can be found on github: https://github.com/KingCounty-Science/nla_explore
$ quarto --version
1.2.335

I did update RStudio and Quarto to version 1.5.57 but then I could no longer publish to Quarto getting the message: "ERROR: The specified path (.) is not a website, manuscript or book project so cannot be published."

I've attached images of the rendered figure and the html file with the blank space where the figure should be.
![Screenshot 2024-10-23 162252](https://github.com/user-attachments/assets/9acc5176-edf3-4cf3-ab26-c7b25f0cb376)
![Screenshot 2024-10-23 162230](https://github.com/user-attachments/assets/db6e058e-96f9-4375-91f8-f1bb3a425bd8)

The qmd file looks like this:

````qmd
---
title: "Compare King Co Lakes With National Lakes Assessment Data"
author: "Curtis DeGasperi"
format: html
server: shiny
date: today
---

## Using Puget Lowland National Lakes Assessment (NLA) data to put King Co. stewardship lakes in context

```{r}
#| echo: false
#| warning: false

# Load data for plotting
suppressPackageStartupMessages(library(tidyverse, quietly = T))
suppressWarnings(library(lubridate))
suppressWarnings(library(plotly))

### King Co and NLA data - Chla and TP

nla_ecdf.chla <- readRDS('nla_ecf.chla.RDS')
nla_ecdf.tp <- readRDS('nla_ecf.tp.RDS')

###
```

```{r}
#| echo: false
#| warning: false

# library(shiny)

nla_ecdf.chla <- readRDS('nla_ecf.chla.RDS')
nla_ecdf.tp <- readRDS('nla_ecf.tp.RDS')

lakes <- sort(unique(nla_ecdf.chla$SiteName[!nla_ecdf.chla$Type=='NLA']))
selectInput("SiteName", "Choose Lake:", lakes, selected = "Welcome")

###
```

```{r}

mainPanel(plotlyOutput('graph'))

###
```

```{r}
#| echo: false
#| warning: false
#| context: server

suppressWarnings(library(plotly))

nla_ecdf.chla <- readRDS('nla_ecf.chla.RDS')
nla_ecdf.tp <- readRDS('nla_ecf.tp.RDS')

a1 <- reactive({
  # m <- nla_ecdf.chla[nla_ecdf.chla$SiteName=='Star', ]
  m <- nla_ecdf.chla[nla_ecdf.chla$SiteName==input$SiteName, ]
  a <- list(
  x = log10(m$Value),
  y = m$ecdf,
  text = m$SiteName,
  xref = "x",
  yref = "y",
  showarrow = TRUE,
  arrowhead = 7,
  ax = 70,
  ay = 0
)
  return(a)
})

a2 <- reactive({
  # m <- nla_ecdf.tp[nla_ecdf.tp$SiteName=='Star', ]
  m <- nla_ecdf.tp[nla_ecdf.tp$SiteName==input$SiteName, ]
  a <- list(
  x = m$Value,
  y = m$ecdf,
  text = m$SiteName,
  xref = "x",
  yref = "y",
  showarrow = TRUE,
  arrowhead = 7,
  ax = 70,
  ay = 0
)
return(a)
})

  output$graph <- renderPlotly({
    fig1 <- plot_ly(nla_ecdf.chla, x = ~log10(Value), y = ~ecdf, color = ~Type, showlegend = FALSE, type = 'scatter', mode = 'markers', hoverinfo = 'text',
                text = ~paste('</br> Value: ', round(Value,1),
                              '</br> Fn(Value): ', round(ecdf,2),
                              '</br> Lake: ', SiteName))
fig1 <- fig1 %>% layout(annotations = a1(), title = "Chlorophyll a (µg/L)", xaxis = list(title = "log10[Chlorophyll a (µg/L)]", size = 14))

fig2 <- plot_ly(nla_ecdf.tp, x = ~Value, y = ~ecdf, color = ~Type, type = 'scatter', mode = 'markers', hoverinfo = 'text',
                text = ~paste('</br> Value: ', round(Value,1),
                              '</br> Fn(Value): ', round(ecdf,2),
                              '</br> Lake: ', SiteName))
fig2 <- fig2 %>% layout(annotations = a2(), title = "Compare King Co Lakes with National Lakes Assessment Data", xaxis = list(title = "Total Phosphorus (µg/L)", size = 14))

    subplot(fig1,fig2,shareY=TRUE, titleX = TRUE, titleY = TRUE)

  })
```

````

**Comment by mcanouil** [comment]
Could you share a small example of the issue?
Could you also properly format your post using code blocks for code and terminal outputs? Thanks.
If your code contains code blocks, you need to enclose it using more backticks, *i.e.*, usually four ` ```` `.
See <https://quarto.org/bug-reports.html#formatting-make-githubs-markdown-work-for-us>.

You should not use inline R expression, especially because it's not necessary https://quarto.org/docs/reference/dates.html

Shiny documents requires some special treatments, did you follow: https://quarto.org/docs/interactive/shiny/running.html

Finally, you are using Quarto 1.2. Current stable release is 1.5. Please upgrade.

**Reply by curtisld** [reply]
Not sure how to make it any smaller. I tried to fix the formatting...replaced the inline r code with 'today'....and update Quarto to 1.5.57 but that resulted in the inability to publish to Quarto. I'm getting this error:  "ERROR: The specified path (.) is not a website, manuscript or book project so cannot be published."

I didn't have any trouble publishing this to shinyapps.io, but I suppose there could be something missing that Quarto requires?

**Reply by mcanouil** [reply]
The issue is not only that's a bit complex/big but it's not reproducible.
We don't have your RDS files.

**Comment by cscheid** [comment]
```
$ quarto --version
1.2.335
```

This version is about a year out of date. Please install a current version of Quarto first. Thanks!

**Reply by curtisld** [reply]
I update quarto and now I can't publish. Getting the message: "ERROR: The specified path (.) is not a website, manuscript or book project so cannot be published."

**Reply by cscheid** [reply]
That looks to be a separate issue. Can you start a separate discussion? Thanks!

**Reply by curtisld** [reply]
I might, but I haven't really tried to solve that on my own yet. I still have the same problem with the graph not appearing in the html file produced when I render my .qmd file. I realize now that the issue isn't with deploying the html file to Quarto pub. So now with the latest version of Quarto, updated packages, and even the latest versions of R and RStudio. Render generates the expected result with my interactive graph, but the html file produced does not contain the graph.

**Comment by mcanouil** [comment, marked as answer]
Here is a fully reproducible example:

<table>
<tr><th>Input</th><th>Output</th></tr>
<tr><td>

````qmd
---
title: "Issue 11175"
format: html
server: shiny
---

```{r}
shiny::sliderInput("bins", "Number of bins:", min = 1, max = 50, value = 30)
plotly::plotlyOutput("distPlot")
```

```{r}
#| context: server
data("diamonds", package = "ggplot2")
output$distPlot <- plotly::renderPlotly({
  plotly::plot_ly(diamonds, x = ~price, y = ~ interaction(clarity, cut)) |>
    plotly::add_boxplot(color = ~clarity) |>
    plotly::layout(yaxis = list(title = ""))
})
```
````

</td><td>

<img width="975" alt="image" src="https://github.com/user-attachments/assets/8263bdb3-299d-4d11-9e05-a8f5932cdc27">

</td></tr>
</table>


Because `server: shiny` it means you need a running R server to run your Shiny App.
Quarto Pub has no server ability, thus the code cell `context: server` is not executed leading to nothing displayed as expected.
An interactive document is considered a Shiny App, so you need to follow https://quarto.org/docs/interactive/shiny/running.html and deploy on a service that has server ability such as shinyapps.io

---

Using a static document:


<table>
<tr><th>Input</th><th>Output</th></tr>
<tr><td>

````qmd
---
title: "Issue 11175"
format: html
---

```{r}
data("diamonds", package = "ggplot2")
plotly::plot_ly(diamonds, x = ~price, y = ~ interaction(clarity, cut)) |>
  plotly::add_boxplot(color = ~clarity) |>
  plotly::layout(yaxis = list(title = ""))
```
````

</td><td>

<img width="1084" alt="image" src="https://github.com/user-attachments/assets/4f3185de-77be-439a-957a-d1929c658279">

</td></tr>
</table>

**Reply by curtisld** [reply]
Ah. Perhaps this is clearly stated somewhere in the Quarto help documents. If not, that seems worth mentioning somewhere within the available shiny interactive examples or where publishing is described. I did something similar to your example...copied one of the Quarto shiny application examples and confirmed that those applications also don't result in an interactive graph in the html files produced. I suspect if I'd done that at the beginning I would have gotten more quickly to the crux of the issue. Apologies for not doing that initially.  Thanks!
